<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

*   Acquire data in various ways
*   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="https://#data_acquisition">Data Acquisition</a>
    <li><a href="https://#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Automobile Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

<ul>
    <li>Data source: <a href="https://1824251045.rsc.cdn77.org/web/algohouse/data/XRPBUSD_trades_1m.csv" target="_blank">https://1824251045.rsc.cdn77.org/web/algohouse/data/XRPBUSD_trades_1m.csv</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


You are running the lab in your  browser, so we will install the libraries using `piplite`.


In [ ]:
#you are running the lab in your  browser, so we will install the libraries using ``piplite``
import piplite
import micropip
await piplite.install(['pandas'])
await piplite.install(['matplotlib'])
await piplite.install(['scipy'])
await piplite.install(['seaborn'])
await micropip.install(['ipywidgets'],keep_going=True)
await micropip.install(['tqdm'],keep_going=True)

If you run the lab locally using Anaconda, you can load the correct library and versions by uncommenting the following:


In [ ]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

In [1]:
# import pandas library
import pandas as pd
import numpy as np

This function will download the dataset into your browser


In [2]:
#This function will download the dataset into your browser 

from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

You can also assign the dataset to any variable you create.

</p>


In [3]:
path = "https://1824251045.rsc.cdn77.org/web/algohouse/data/XRPBUSD_trades_1m.csv"

You will need to download the dataset; if you are running locally, please comment out the following:


In [4]:
# you will need to download the dataset; if you are running locally, please comment out the following 
await download(path, "XRPBUSD_trades_1m.csv")
path="XRPBUSD_trades_1m.csv"

This dataset was hosted on IBM Cloud object. Click <a href="https://cocl.us/DA101EN_object_storage?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01">HERE</a> for free storage.


In [5]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"
df = pd.read_csv(path)

In finance you sometimes need to use different numbers of decimal places. For ease of reading, let's specify the value of the precision parameter equal to 3 to display three decimal signs (instead of 6 as default).

In [6]:
pd.set_option("display.precision", 3)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [7]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


Unnamed: 0                   ts   open   high    low  close    volume  \
0           0  2022-11-11 14:38:00  0.370  0.370  0.368  0.368  165726.4   
1           1  2022-11-11 14:39:00  0.368  0.369  0.367  0.368  235942.4   
2           2  2022-11-11 14:40:00  0.368  0.368  0.367  0.367  475118.5   
3           3  2022-11-11 14:41:00  0.367  0.369  0.367  0.368  420195.5   
4           4  2022-11-11 14:42:00  0.368  0.369  0.367  0.367  199091.6   

   rec_count  avg_price  
0         47      0.369  
1         49      0.368  
2         46      0.367  
3         53      0.368  
4         42      0.368

As there is no text (i.e. column title) on the first cell of the csv file, the resulting dataframe first column is given the name <code>Unnamed:0</code>. We should fix this issue.

By specifying an <code>index_col=[0]</code> argument to <code>read_csv()</code> function we tell pandas that the first column in the csv file is the index for the dataframe. As follows the undesired column <code>Unnamed:0</code> will disappear.


In [8]:
df = pd.read_csv(path, index_col=0)
df.head(5)

ts   open   high    low  close    volume  rec_count  \
0  2022-11-11 14:38:00  0.370  0.370  0.368  0.368  165726.4         47   
1  2022-11-11 14:39:00  0.368  0.369  0.367  0.368  235942.4         49   
2  2022-11-11 14:40:00  0.368  0.368  0.367  0.367  475118.5         46   
3  2022-11-11 14:41:00  0.367  0.369  0.367  0.368  420195.5         53   
4  2022-11-11 14:42:00  0.368  0.369  0.367  0.367  199091.6         42   

   avg_price  
0      0.369  
1      0.368  
2      0.367  
3      0.368  
4      0.368

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [24]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Add Headers</h3>
<p>
Take a look at our dataset. If we do not specify the header of our dataset by passing an argument <code>headers = None</code> inside the <code>read_csv()</code> method, Pandas automatically sets it with an integer starting from 0.
</p>
<p>
To better describe our data, it is a best practice to introduce a header. 
Information of dataset we are using is available at: <a href="https://1824251045.rsc.cdn77.org/web/algohouse/data/XRPBUSD_trades_1m.csv" target="_blank">https://1824251045.rsc.cdn77.org/web/algohouse/data/XRPBUSD_trades_1m.csv</a>.
</p>
<p>
Thus, if headers are absent, we have to add them manually.
</p>
<p>
First, we create a list "headers" that include all column names in order.
Then, we use <code>dataframe.columns = headers</code> to replace the headers with the list we created.
</p>


In [15]:
# create headers list
headers = ["Ts", "Open", "High", "Low", "Close", "Volume", "Rec_count", "Avg_price"]
print("headers\n", headers)

headers
 ['Ts', 'Open', 'High', 'Low', 'Close', 'Volume', 'Rec_count', 'Avg_price']


We replace headers and recheck our dataframe:


In [16]:
df.columns = headers
df.head(10)

Ts   Open   High    Low  Close     Volume  Rec_count  \
0  2022-11-11 14:38:00  0.370  0.370  0.368  0.368  1.657e+05         47   
1  2022-11-11 14:39:00  0.368  0.369  0.367  0.368  2.359e+05         49   
2  2022-11-11 14:40:00  0.368  0.368  0.367  0.367  4.751e+05         46   
3  2022-11-11 14:41:00  0.367  0.369  0.367  0.368  4.202e+05         53   
4  2022-11-11 14:42:00  0.368  0.369  0.367  0.367  1.991e+05         42   
5  2022-11-11 14:43:00  0.367  0.368  0.367  0.368  2.034e+05         31   
6  2022-11-11 14:44:00  0.368  0.368  0.368  0.368  1.530e+05         29   
7  2022-11-11 14:45:00  0.369  0.375  0.369  0.374  8.787e+05         74   
8  2022-11-11 14:46:00  0.375  0.377  0.373  0.377  1.281e+06         73   
9  2022-11-11 14:47:00  0.377  0.385  0.377  0.383  2.174e+06         95   

   Avg_price  
0      0.369  
1      0.368  
2      0.367  
3      0.368  
4      0.368  
5      0.368  
6      0.368  
7      0.372  
8      0.375  
9      0.382

 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [17]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>XRPBUSD_trades_1m.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.

</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas to return series with the data type of each column of dataframe <code>.dtypes</code> is used.
</p>


In [18]:
# check the data type of data frame "df" by .dtypes
print(df.dtypes)

Ts            object
Open         float64
High         float64
Low          float64
Close        float64
Volume       float64
Rec_count      int64
Avg_price    float64
dtype: object


<p>
As shown above, it is clear to see that the data type of "open", "high", "low", "close", "avg_price", and "volume" are <code>float64</code>, "ts" is <code>object</code>, and "rec_count" is <code>int64</code>, etc.
</p>
<p>
These data types can be changed. For instance, the "ts" field should be converted to <b>datatime64</b> type as it represents the date and time. We will learn how to accomplish this in a later module.
</p>

<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [19]:
# get a statistical summary of each column of dataframe using .describe()
df.describe()

Open       High        Low      Close     Volume  Rec_count  \
count  66911.000  66911.000  66911.000  66911.000  6.691e+04  66911.000   
mean       0.374      0.374      0.374      0.374  1.206e+05     13.393   
std        0.019      0.019      0.019      0.019  2.417e+05     11.382   
min        0.321      0.322      0.321      0.321  1.000e-01      1.000   
25%        0.354      0.354      0.353      0.354  2.061e+04      6.000   
50%        0.379      0.379      0.379      0.379  5.722e+04     10.000   
75%        0.389      0.389      0.389      0.389  1.335e+05     17.000   
max        0.417      0.418      0.416      0.417  1.437e+07    130.000   

       Avg_price  
count  66911.000  
mean       0.374  
std        0.019  
min        0.321  
25%        0.354  
50%        0.379  
75%        0.389  
max        0.417

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "rec_count" has 66911 counts, the mean value of this column is 13.393, the standard deviation is 11.382, the minimum value is 1, 25th percentile is 6, 50th percentile is 10, 75th percentile is 17, and the maximum value is 130.<br>

However, what if we would also like to check all the columns including those that are of type *object*? <br><br>
You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.
</p>


In [20]:
# describe all the columns in "df" 
df.describe(include = "all")

Ts       Open       High        Low      Close  \
count                 66911  66911.000  66911.000  66911.000  66911.000   
unique                66911        NaN        NaN        NaN        NaN   
top     2022-11-11 14:38:00        NaN        NaN        NaN        NaN   
freq                      1        NaN        NaN        NaN        NaN   
mean                    NaN      0.374      0.374      0.374      0.374   
std                     NaN      0.019      0.019      0.019      0.019   
min                     NaN      0.321      0.322      0.321      0.321   
25%                     NaN      0.354      0.354      0.353      0.354   
50%                     NaN      0.379      0.379      0.379      0.379   
75%                     NaN      0.389      0.389      0.389      0.389   
max                     NaN      0.417      0.418      0.416      0.417   

           Volume  Rec_count  Avg_price  
count   6.691e+04  66911.000  66911.000  
unique        NaN        NaN        NaN  
top           NaN        NaN        NaN  
freq          NaN        NaN        NaN  
mean    1.206e+05     13.393      0.374  
std     2.417e+05     11.382      0.019  
min     1.000e-01      1.000      0.321  
25%     2.061e+04      6.000      0.354  
50%     5.722e+04     10.000      0.379  
75%     1.335e+05     17.000      0.389  
max     1.437e+07    130.000      0.417

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>

</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'Rec_count' and 'Avg_price'.

</div>


In [23]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
df[['Rec_count', 'Avg_price']].describe()
```

</details>


<h2>Info</h2>
Another method you can use to check your dataset is:


It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [13]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 66911 entries, 0 to 66910
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Ts         66911 non-null  object 
 1   Open       66911 non-null  float64
 2   High       66911 non-null  float64
 3   Low        66911 non-null  float64
 4   Close      66911 non-null  float64
 5   Volume     66911 non-null  float64
 6   Rec_count  66911 non-null  int64  
 7   Avg_price  66911 non-null  float64
dtypes: float64(6), int64(1), object(1)
memory usage: 4.3+ MB


<h1>Excellent! You have just completed the  Introduction Notebook!</h1>


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Fiorella Wenver</a>

<a href="https:// https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                         |
| ----------------- | ------- | ---------- | ---------------------------------------------------------- |
| 2022-08-23        | 2.4     | Malika     | Import micropip added and parameter for ipwidgets and tqdm |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                                     |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'                       |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe                   |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab                         |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
